norm: rule to simplify const comparison with ANY values

dils2k opened the pull request:
Given a constant value compared using the ANY or
SOME operators, we check for constants
in the right hand side that can be used to
simplify the comparison to True. For example,
the expression: 0.1 < ANY (c0, 0.4) can now be simplified since 0.1 < 0.4.

Fixes #132328
Release note: None
Epic: CRDB-42979

DrewKimball reviewed (commented):
This is looking good! One thing I believe your logic is missing (and which should be tested) is proper handling for NULL values. Here's an example:
```
root@localhost:26257/defaultdb> select 1 = any (2, 3, null);
  ?column?
------------
    NULL
(1 row)
```
So the rule should be:
* If any comparison folds to `true`, return `true`.
* Otherwise, if all operands are constant and any comparison folds to `NULL`, return `NULL`.
* Otherwise, if all operands are constant, return `false`.

> @DrewKimball reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OcCVJVl7enZVMiHbCEo:bt6h3ka)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bghal, @mgartner, and @rytaft)

___
*[`pkg/sql/opt/norm/fold_constants_funcs.go` line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-OcCVedF8FU195hIHTJY:-OcCVedF8FU195hIHTJZ:b3b95lq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d05b5f08581f3c985225fb0f4534e2166ca438d/pkg/sql/opt/norm/fold_constants_funcs.go#L210)):*
> ```Go
> 	if !ok {
> 		return nil, false
> 	}
> ```

nit: It would be nice to pull out these first two checks into the optgen rule.


<!-- Sent from Reviewable.io -->

dils2k commented:
@DrewKimball updated the rule

dils2k commented:
@DrewKimball a kind ping in case my PR got lost 🙏🏻

DrewKimball reviewed (commented):
Sorry for the late response. The rule looks good to me! The only thing missing at this point is a couple "no-op" tests demonstrating interesting cases where the rule _shouldn't_ fire. For example, cases where one constant comparison evaluates to `NULL` or `false`, but there's a non-constant expression. You can verify that the rule doesn't fire using:
```
norm expect-not=FoldComparisonWithAny
```

> @DrewKimball reviewed 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OdKHgCF11pMEOT4VibJ:b-cmokgu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bghal, @dils2k, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

dils2k commented:
@DrewKimball added more tests, including ones with `expect-not`

DrewKimball reviewed (approved):
Nice work! This <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @DrewKimball reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OdUfPk48uTxncoIf9oR:b-wbxi8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @bghal, @dils2k, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

bghal commented on pkg/sql/opt/norm/fold_constants_funcs.go:
```diff
@@ -193,6 +195,37 @@ func (c *CustomFuncs) IsConstValueOrGroupOfConstValues(input opt.ScalarExpr) boo
 	return memo.CanExtractConstDatum(input)
 }
 
+func (c *CustomFuncs) FoldComparisonWithAny(cmp opt.Operator, left, right opt.ScalarExpr) (opt.ScalarExpr, bool) {
```
Would you mind making this comment consistent with the other comments?
<!-- Reviewable comment -OdsSnIK6pTVlvcNhS2Q:b-c437rf -->
<!-- Sent from Reviewable.io -->

bghal replied:
This should be more similar to the comments on the [other](https://github.com/search?q=repo%3Acockroachdb%2Fcockroach+%22%28c+*CustomFuncs%29+Fold%22&type=code) fold evaluations. Preferably it'd be a higher-view rather than the specific types returned.
<!-- Reviewable comment -OeHXB8b3xCVrB1YubV2:bek6a7w -->
<!-- Sent from Reviewable.io -->

bghal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dils2k, @DrewKimball, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

bghal reviewed (commented):
@DrewKimball Claude seems to have his objections you mind looking at that workflow?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OdsTH0sDCffGW1ZF2GZ:bhv1bvn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dils2k, @DrewKimball, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
>Claude seems to have his objections you mind looking at that workflow?

I think the claude workflow just doesn't know how to handle an external contributor. I think it's fine to ignore that and just merge.

bghal reviewed (changes requested):
Nearly there. More on comments and there's an open nit to address or close.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OeHcJdgCz4en-ijITm_:b11o3aj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dils2k, @DrewKimball, @mgartner, and @rytaft)

___
*[`pkg/sql/opt/norm/rules/fold_constants.opt` line 104 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-OeHY6Du4JNStQWRDkwZ:-OeHY6Du4JNStQWRDkw_:b-b286ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/46f923eb647bc91cf3981c4c54b57bf642b57b46/pkg/sql/opt/norm/rules/fold_constants.opt#L104)):*
> ```
> $result
> 
> [FoldComparisonWithAny, Normalize]
> ```

Does this need a comment similar to the other rules?
___
*[`pkg/sql/opt/norm/fold_constants_funcs.go` line 206 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-OeHc0CV9y7SDDIwuEE_:-OeHc0CV9y7SDDIwuEEa:b-fudwj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/46f923eb647bc91cf3981c4c54b57bf642b57b46/pkg/sql/opt/norm/fold_constants_funcs.go#L206)):*
> ```Go
> func (c *CustomFuncs) FoldComparisonWithAny(
> 	cmp opt.Operator, left, right opt.ScalarExpr,
> ) (opt.ScalarExpr, bool) {
> ```

Would it make sense to use the pattern of named returns?



_Code snippet:_
```Go
) (_ opt.ScalarExpr, ok bool) {
```


<!-- Sent from Reviewable.io -->

dils2k committed:
norm: rule to simplify const comparison with ANY values

Given a constant value compared using the ANY or
SOME operators, we check for constants
in the right hand side that can be used to
simplify the comparison to True. For example,
the expression: 0.1 < ANY (c0, 0.4) can now be simplified since 0.1 < 0.4.

Fixes #132328
Release note: None
Epic: CRDB-42979

dils2k commented:
@bghal updated the comments, please let me know if they are better this time 🙏🏻

bghal reviewed (commented):
Looks good thanks for your contribution.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OeRtDx9BE3-sxU04jvS:b-jt71o3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dils2k, @DrewKimball, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

bghal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @bghal reviewed 1 of 1 files at r5, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155212#-:-OeRtNQw0yyanaaY2rT1:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dils2k, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->
